### ISSUE TITLE
16F5F not centered in head position

### ISSUE THREAD

**Issue by kienwt** [silnrsi/font-shimenkan#50]
Raw text
[raw-not-centered-head-position.txt](https://github.com/silnrsi/font-shimenkan/files/3221521/raw-not-centered-head-position.txt)

As displayed in LibreOffice
![raw-not-centered-head-position](https://user-images.githubusercontent.com/18396131/58390905-193da280-8066-11e9-83ae-8eba8b179e1c.PNG)

**Comment by cheuk879** [comment]
Per build 222, all but [hmdd] still render it off-centre:
![plrd-test-centring-16f5f](https://user-images.githubusercontent.com/10625598/61917416-f0207d80-af01-11e9-9d8d-526acd6504ff.png)

**Comment by mhosken** [comment]
This behaviour was added in response to issue #43. Which way would you like us to jump?

**Comment by cheuk879** [comment]
Issue #43 is an adjustment that only applies to 16F63..16F64.  The latter is only used by [hmd], & the former is not currently used by any group, but [hmd] requested that it be treated like the latter in case it would be used later.

The current issue here is a defect in the position of 16F5F that affects all langs but [hmdd].  Centring is the correct behaviour, but per build 233 it is still off-centre as above.

**Comment by mhosken** [comment]
Is this what you want? 
![image](https://user-images.githubusercontent.com/5687149/62098344-181e3100-b2b4-11e9-9eed-caddf554d5aa.png) committed dcf4b19a

**Comment by cheuk879** [comment]
Yes.  16F5F confirmed to be centred per build 242.  However, 16F63..16F64 are clashing with asp:
![plrd-test-16f63-16f64-err](https://user-images.githubusercontent.com/10625598/62672780-d561f600-b950-11e9-9a24-8b5d08f8a09a.png)

**Comment by cheuk879** [comment]
No clashing per Build 244:
![plrd-test-16f63-16f64](https://user-images.githubusercontent.com/10625598/62758310-df5c2580-ba32-11e9-829b-75bb68e3cda4.png)
However, in the rightmost column, V is almost touching the asp & can use just a little bit shift to the right.  And for those in red, the asp is a bit too far away from the base char & can use a bit more shift to the left.

**Comment by jvgaultney** [comment]
I'd agree that something odd is happening with the position of the asp, and the problem is in the OT code.

**Comment by mhosken** [comment]
Dear all,

The aspiration is set first. So any adjustment is to the vowel, which I
guess aligns it's centre with the RHS of the aspiration. If you give me an
AP on the vowel to say where the RHS of the aspiration goes, I would use
that. If we want to move the aspiration too, then we have to change the
maths. Which we can do. But I suggest this is post 1.0

GB,
Martin

**Comment by jvgaultney** [comment]
Apologies - this does seem to be related to glyph widths (and APs). The _HD AP on the .mark vowels should be attaching to the HD AP on the asp. That may actually be happening correctly and I just need to move the APs around a bit.

The inconsistency in the asp position related to the cons is related to glyph width (which in turn defines certain AP locations). There's normally more width for letters that have a straight stem on the right hand side of letter (such as 1E) than those that have a lot of white space on the lower right (such as 0A). If I make the right side of them mathematically identical then it will visually look inconsistent when at text size, and certain vowels at F will look disconnected. However I may have the difference too great for some weights. It's a really tricky balance, and I couldn't fine tune it until all the many many issues around dots, asp, flush right, etc. were finalized.

I will try to fine-tune both the asp+vowel APs and the glyphs widths, however that will very likely need to wait until after 1.0 because we're running out of time. (Remember that unless we release v1.0 very soon we'll miss the window time I have available and it will all have to wait until November or so.)

**Comment by cheuk879** [comment]
If timing is an issue then yes pls focus on releasing 1.0 first.

**Comment by jvgaultney** [comment]
Actually the main problems here are OT related. Here is a composite image from an updated issues.htxt test files at three weights in build 275. Note that there are no space characters in these strings:

![image](https://user-images.githubusercontent.com/4226750/63174417-d4f4e980-c039-11e9-9060-8862de6b0a7b.png)

There are a number of problems here.

1) Space is being added to the right side of any combination that involves u16F63 or u16F64 at head position, even without an asp.

2) More space is then added when adding asp to those combos (but only if u16F63 or u16F64 are present).

3) The alignment of asp and V is inconsistent - they collide even more with u16F28 than with others. However u16F28 isn't likely to be the cause, as it's working fine with only asp. (Note that the right sidebearing of u16F28 is a little larger than other letters, but only a tiny amount - much smaller than the collision difference.

I can't figure out which AP on asp is being used to position the V in these cases. The head alignment of these vowels at head position in general could probably still be improved by adjusting APs, but I can't do that until these odd bugs are fixed.



(note that the extraspace

**Comment by mhosken** [comment]
The HD AP on a vowel is used, and is aligned to the righthand of the asp bounding box. A bug with the U+16F28 (I shape) has been fixed that should bring the asp back and spacing fixed, I hope.

**Comment by jvgaultney** [comment]
Spacing bugs confirmed to be fixed in build 279. Further testing and improvement of alignments for these overhanging vowels will need to wait until after 1.0.

**Comment by jvgaultney** [comment]
A spacing problem has returned as a result of the Windows fix for #82. Spacing is being added when an asp that is positioned after is followed by V@H. Affects [ygp] (and presumably [sfm, yna] too), but is OK for other languages for which the asp is tucked.

[ygp] (ygp_issues.htxt):

![image](https://user-images.githubusercontent.com/4226750/63511526-584e8900-c4d9-11e9-915e-3af115b285d5.png)

default (issues.htxt):

![image](https://user-images.githubusercontent.com/4226750/63511554-67353b80-c4d9-11e9-8b91-56a46e1426e0.png)

If possible it would be good to fix this for 1.0.

**Comment by jvgaultney** [comment]
Extra space now confirmed to be removed in build 292:

![image](https://user-images.githubusercontent.com/4226750/63523260-b6876600-c4f1-11e9-91d1-50357162cc52.png)

However spacing remains inconsistent with vs. without asp. Oddly with asp is the 'correct' spacing according to the glyph metrics. That needs fixing, but then the without asp will need to be reviewed. With and without should have identical right sidebearings.

These adjustments can wait until after 1.0.
